### PR TITLE
Bug fix to allow recovery of the DKG to be able to gen priv keys

### PR DESCRIPTION
**Pull request by n-hutton**
Save the dkg information so that if a node crashes after submitting dry run keys, when it replays the blocks on a restart the aeon keys it outpus will be able to sign.

This avoids/minimises the scenario where threshold validators crash at that point in the DKG causing the network to stall

### CODE REVIEW

File: beacon/aeon_details.go [hunk 1]
```diff
@@ -185,6 +197,45 @@ func loadAeonDetails(aeonDetailsFile *AeonDetailsFile, validators *types.Validat
 	return aeonDetails, nil
 }
 
+// Add aon aeon to the file, keeping a max of N aeons in the file (reading
+// from the file and appending if neccessary)
+func updateFileAeons(filePath string, max int, aeons ...*aeonDetails) {
```
Comment (n-hutton) [inline beacon/aeon_details.go:202]: Function to ensure N (2) aeons are in the old aeons file

File: beacon/dkg.go [hunk 2]
```diff
@@ -74,13 +91,15 @@ func newState(durMultiplier int64, entry func(), exit func() bool, transition fu
...
+	dkgIteration int64
+
 	service.BaseService
 	mtx sync.RWMutex
 
 	config       *cfg.BeaconConfig
-	chainID      string
-	dkgID        int64
-	dkgIteration int64
+	baseConfig   *cfg.BaseConfig
```
Comment (n-hutton) [inline beacon/dkg.go:102]: add baseconfig to give file location for dkg file

File: beacon/dkg.go [hunk 3]
```diff
@@ -473,16 +517,32 @@ func (dkg *DistributedKeyGeneration) checkTransition(blockHeight int64) {
...
+		dkg.proceedToNextState(currentState+1, true, blockHeight)
 	}
 }
 
 func (dkg *DistributedKeyGeneration) proceedToNextState(nextState dkgState, runOnEntry bool, blockHeight int64) {
+
+	// If the state we are going into is the first one, we can see if it is possible to load a DKG which has crashed.
+	// Otherwise, we save our dkg details
+	if dkg.enableRecovery {
+		if nextState == waitForEncryptionKeys {
```
Comment (n-hutton) [inline beacon/dkg.go:534]: Load the beacons service on the first state, save it on the last states

File: mcl_cpp/base_dkg.hpp [hunk 4]
```diff
@@ -228,19 +307,30 @@ class BaseDkg {
 
   // Temporary variables in DKG
   PrivateKey xprime_i_;
-  std::vector<std::vector<PrivateKey> > s_ij_, sprime_ij_;
+  std::vector<std::vector<PrivateKey>> s_ij_, sprime_ij_;
   std::vector<std::vector<VerificationKey>> C_ik_;
   std::vector<std::vector<VerificationKey>> A_ik_;
   std::vector<std::vector<VerificationKey>> secret_commitments_;
 
+  // Save the coefficients after generating so they can be recovered in the case of a crash
```
Comment (n-hutton) [inline mcl_cpp/base_dkg.hpp:315]: save the coefficients to the base class

File: beacon/dkg.go [hunk 5]
```diff
@@ -133,6 +142,7 @@ func NewDistributedKeyGeneration(beaconConfig *cfg.BeaconConfig, chain string,
 		entropyParams:        entropyParams,
 		threshold:            dkgThreshold,
 		startHeight:          validatorHeight,
+		enableRecovery:       true,
```
Comment (jinmannwong) [inline beacon/dkg.go:145]: Is this just so we can turn off recovery for tests?

File: beacon/dkg.go [hunk 6]
```diff
@@ -92,6 +98,7 @@ type DistributedKeyGeneration struct {
 	stateDuration   int64
 	aeonKeys        *aeonDetails
 	onFailState     func(int64)
+	enableRecovery  bool
```
Comment (jinmannwong) [inline beacon/dkg.go:101]: Is this so we can turn off recovery for tests?

**Review by jinmannwong — APPROVED** [review]

File: beacon/aeon_details.go (hunk 1 above)
Comment (jinmannwong) [reply beacon/aeon_details.go]: Wasn't this change already merged?

File: beacon/dkg.go (hunk 6 above)
Comment (n-hutton) [reply beacon/dkg.go:101]: Yup

File: beacon/aeon_details.go (hunk 1 above)
Comment (n-hutton) [reply beacon/aeon_details.go]: Ah, I realised that I didn't close that PR at the time I made the comment, yes it was merged previously and thus was not in this PR

**Pull request merged**